Remove GetDepthInMainChain dependency on locked chain interface

ariard opened the pull request:
Work starter to remove Chain::Lock interface by adding m_last_block_processed_height in CWallet and m_block_height in CMerkleTx to avoid GetDepthInMainChain having to keep a lock . Once this one done, it should ease work to wipe out more cs_main locks from wallet code.

I think it's ready for a first round of review before to get further.

- `BlockUntilSyncedToCurrent` : restrain isPotentialTip to isTip because we want to be sure that wallet see BlockDisconnected callbacks if its height differs from the Chain one. It means during a reorg, an RPC could return before the BlockDisconnected callback had been triggered. This could cause a tx that had been included in the disconnected block to be displayed as confirmed, for example.

~~- `AbandonTransaction` : in case of conflicted tx (nIndex = -1), we set its m_block_height to the one of conflicting blocks, but if this height is superior to CWallet::m_last_block_processed_height, that means tx isn't conflicted anymore so we return 0 as tx is again unconfirmed~~ After #16624, we instead rely on Confirmation. 

~~- `AddToWalletIfInvolvingMe`: in case of block disconnected, transactions are added to mempool again, so we need to replace old txn in `mapWallet` with a height set to zero so we remove check on block_hash.IsNull~~ Already done in #16624

Commit:
Add block_height field in struct Confirmation

At wallet loading, we rely on chain state querying to retrieve
height of txn, to do so we ensure that lock order is respected
between cs_main and cs_wallet.

If wallet loaded is the wallet-tool one, all wallet txn will
show up with a height of zero. It doesn't matter as confirmation
height is not used by wallet-tool.

Reorder arguments and document Confirmation calls to avoid
ambiguity.

Fixes nits left from #16624

Commit:
Add m_last_block_processed_height field in CWallet

At BlockConnected/BlockDisconnected, we rely on height of block
itself to know current height of wallet

Commit:
Use CWallet::m_last_block_processed_height in GetDepthInMainChain

Avoid to lock chain to query state thanks to tracking last block
height in CWallet.

Commit:
Remove locked_chain from GetDepthInMainChain and its callers

We don't remove yet Chain locks as we need to preserve lock
order with CWallet one until swapping at once to avoid
deadlock failures (spotted by --enable-debug)

Commit:
Remove getBlockDepth method from Chain::interface

Pass conflicting height in CWallet::MarkConflicted

fanquake commented:
Added to the "Chasing Concept ACK" board. @ryanofsky maybe you could give an initial Concept/Approach ACK/NACK ?

ariard commented:
Yes would be awesome to get a Concept ACK on the approach followed but if it not the best one, I'm eager to rework the PR in depth! Going to rebase/fix tests failure soon

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -4448,10 +4448,18 @@ void CMerkleTx::SetMerkleBranch(const uint256& block_hash, int posInBlock, int b
+        return 0;
+
     if (hashUnset())
         return 0;
```
In commit "Use CWallet::m_last_block_processed_height in GetDepthInMainChain" (233ae92f2d51333f158810d85d13cf63a1144e6b)

Can we `assert(m_block_height > 0)` after this point? I'm afraid of another bug like the `importprunedfunds` one mentioned earlier leaving m_block_height unset and this function returning bogus values.

ariard replied:
Hmmm, I introduced  ```assert(m_block_height >= -1)``` as we may GetDepthInMainChain on fresh transactions, their ```hashBlock``` being unset, their depth should appear as zero.

ariard replied:
Note : I've seen one Travis failure on feature_reindex.py where the assert failed. I wasn't able to reproduce the failure on my own env. I think that's caused by the slow Travis setup but that's not great, what's your take on it ?

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -4448,10 +4448,18 @@ void CMerkleTx::SetMerkleBranch(const uint256& block_hash, int posInBlock, int b
 
-    return locked_chain.getBlockDepth(hashBlock) * (nIndex == -1 ? -1 : 1);
+    if (nIndex == -1) {
+        // Tx is no more conflicted but it's not more confirmed, so return 0
```
In commit "Use CWallet::m_last_block_processed_height in GetDepthInMainChain" (233ae92f2d51333f158810d85d13cf63a1144e6b)

This looks right but is the logic is kind of obscure. Maybe you could add an explanatory comment like "If nIndex is -1 and the block hash is set, it means the transaction is conflicted, and that m_block_height will be set to the height of the earliest conflicting block, so can return 0 is that block has been disconnected."

I think it'd also be good to replace nested `if(A) { if (B) return; }` with `if (A && B) { return; } `

ariard replied:
Commented and updated logic as suggested on 573ca36

ryanofsky commented on src/wallet/wallet.h:
```diff
@@ -1308,6 +1313,10 @@ class CWallet final : public CCryptoKeyStore, private interfaces::Chain::Notific
     /** Implement lookup of key origin information through wallet key metadata. */
     bool GetKeyOrigin(const CKeyID& keyid, KeyOriginInfo& info) const override;
+
+    int GetLastBlockHeight() const { return m_last_block_processed_height; };
```
In commit "Add m_last_block_processed_height field in CWallet" (0284c72ab07362150fd26c86966ea7c7bc6cfed5)

Can this `assert(m_last_block_processed_height >= 0)`? This would catch usage errors if there are calls to this function before the wallet is initialized.

ariard replied:
Check should be there in b1ae297

ariard replied:
Note : I've seen one Travis failure on feature_reindex.py due to the assert. I wasn't able to reproduce it on my local dev env, I think it's caused by Travis really slow setup, but that's not great, what's your take on it ?

ryanofsky reviewed (commented):
Concept ACK. I left a lot of suggestions, but overall this looks very good, and it's great to get rid of all these cs_main locks.

Commit:
Only return early from BlockUntilSyncedToCurrentChain if current tip
is exact match

In the next commit, we start using BlockConnected/BlockDisconnected
callbacks to establish tx depth, rather than querying the chain
directly.

Currently, BlockUntilSyncedToCurrentChain will return early if
the best block processed by the wallet is a descendant of the node'tip.
That means that in the case of a re-org, it won't wait for the
BlockDisconnected callbacks that have been enqueued during the re-org
but have not yet been triggered in the wallet.

Change BlockUntilSyncedToCurrentChain to only return early if the
wallet's m_last_block_processed matches the tip exactly. This ensures
that there are no BlockDisconnected or BlockConnected callbacks
in-flight.

promag reviewed (commented):
Concept ACK.

ariard commented on src/wallet/wallet.h:
```diff
@@ -383,9 +389,10 @@ class CMerkleTx
         READWRITE(hashBlock);
         READWRITE(vMerkleBranch);
         READWRITE(nIndex);
+        READWRITE(m_block_height);
```
Hmmm filled at startup using ```getBlockHeight```, given we already have ```hashBlock``` ? Do you envision the wallet to always query chain state to succeed its startup ?

ariard commented on src/wallet/wallet.cpp:
```diff
@@ -965,12 +965,19 @@ bool CWallet::AddToWallet(const CWalletTx& wtxIn, bool fFlushOnClose)
+            wtx.m_block_height = wtxIn.m_block_height;
+            fUpdated = true;
+        }
+        if (wtxIn.hashBlock.IsNull() && !wtx.hashBlock.IsNull()) {
```
Updated ```AddToWallet```, hope it's clearer

ariard commented on src/wallet/wallet.cpp:
```diff
@@ -4424,13 +4437,16 @@ CWalletKey::CWalletKey(int64_t nExpires)
     nIndex = posInBlock;
+
+    // set tx height
+    m_block_height = block_height;
```
Oooops, I forgot this one. I've added a call to ```SetMerkleBranch``` in ```importprunedfunds``` but had to change its argument to hask for transaction height too. Seems to me similar to the serialization issue, do we want wallet to ask chain the state of its stored transactions every time, even it should be able to infer it from the blocks already connected ?

ariard commented on src/wallet/wallet.cpp:
```diff
@@ -1090,8 +1097,10 @@ bool CWallet::AddToWalletIfInvolvingMe(const CTransactionRef& ptx, const uint256
             CWalletTx wtx(this, ptx);
 
             // Get merkle branch if transaction was found in a block
```
Should be dropped on 1458ded

ariard commented on src/wallet/wallet.cpp:
```diff
@@ -1090,8 +1097,10 @@ bool CWallet::AddToWalletIfInvolvingMe(const CTransactionRef& ptx, const uint256
             // Get merkle branch if transaction was found in a block
-            if (!block_hash.IsNull())
-                wtx.SetMerkleBranch(block_hash, posInBlock);
+            // Set block_hash to Null, posInBlock to 0, block_height to 0
```
Should be dropped on 1458ded

ariard commented on src/interfaces/chain.h:
```diff
@@ -241,8 +241,7 @@ class Chain
     //! Wait for pending notifications to be processed unless block hash points to the current
-    //! chain tip, or to a possible descendant of the current chain tip that isn't currently
-    //! connected.
+    //! chain tip.
```
Done, see 9c07c16

ariard commented on src/wallet/wallet.cpp:
```diff
@@ -94,10 +94,7 @@ static std::set<CWallet*> g_unloading_wallet_set;
-    // Unregister and delete the wallet right after BlockUntilSyncedToCurrentChain
-    // so that it's in sync with the current chainstate.
     wallet->WalletLogPrintf("Releasing wallet\n");
-    wallet->BlockUntilSyncedToCurrentChain();
```
Splitted between 8a16c67 and 9c07c16

ariard commented on src/wallet/wallet.cpp:
```diff
@@ -1152,6 +1161,8 @@ bool CWallet::AbandonTransaction(interfaces::Chain::Lock& locked_chain, const ui
             // If the orig tx was not in block/mempool, none of its spends can be in mempool
             assert(!wtx.InMempool());
             wtx.nIndex = -1;
+            // Conflicting height is inherited from ancestor tx
```
Okay, I did a second round with the following model in head  (state, triplet<hashBlock, nIndex, block_height>) : NEW<{}, -1, -1>, UNCONFIRMED<{}, 0, 0>, CONFIRMED<block_hash, posInBlock, block_height>, ABANDON<ABANDON_HASH, -1, -1>, CONFLICTED<hashBlock, -1, conflicting_height>.
Do you have one similar, should we noted it somewhere ?

ariard commented on src/wallet/wallet.cpp:
```diff
@@ -1267,11 +1280,11 @@ void CWallet::BlockConnected(const CBlock& block, const std::vector<CTransaction
     m_last_block_processed_height += 1;
     for (const CTransactionRef& ptx : vtxConflicted) {
-        SyncTransaction(ptx, {} /* block hash */, 0 /* position in block */);
+        SyncTransaction(ptx, {} /* block hash */, 0 /* block height */, 0 /* position in block */);
```
Ok, updated on 6289909, maybe I should refine it in its own commit?

ariard commented:
Answered back to @ryanofsky comments on 1458ded, I'm on cleaning last Travis failures

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -4608,36 +4606,48 @@ CWalletKey::CWalletKey(int64_t nExpires)
 {
+    if (pwallet == nullptr)
+        return 0;
+    assert(m_block_height >= -1);
```
This line is introduced in a too-early commit (_Add m_last_block_processed_height field in CWallet_) where the member variable _m_block_height_ doesn't yet exist. The _Add m_last_block_processed_height field in CWallet_ commit doesn't build because the variable is not declared.

ariard replied:
Ooops, bad rebase sorry, should be corrected

jnewbery commented on src/wallet/rpcdump.cpp:
```diff
@@ -375,6 +375,7 @@ UniValue importprunedfunds(const JSONRPCRequest& request)
                 {
                     {"rawtransaction", RPCArg::Type::STR_HEX, RPCArg::Optional::NO, "A raw transaction in hex funding an already-existing address in wallet"},
                     {"txoutproof", RPCArg::Type::STR_HEX, RPCArg::Optional::NO, "The hex output from gettxoutproof that contains the transaction"},
+                    {"height", RPCArg::Type::NUM, RPCArg::Optional::NO, "The height in which transaction has been confirmed"},
```
Is this required? the txoutproof includes a block header, which could be used to look up the height in the chain.

ariard replied:
Hmmm well we could use `getBlockHeight` again in this function but I didn't want because was planning to remove this Chain method after we move ScanForWalletTransactions and its logic on the node side.

Maybe that a bit premature, but IMO this RPC should come with all proofs + height without us having to query the chain..

jnewbery replied:
I don't like this change for a couple of reasons:

- it's a breaking change for clients, so they'll need to reimplement their code
- the RPC doesn't do any checking on the `height` value passed, so it's possible to get the wallet into some inconsistent state by passing a bad `height` value.

jnewbery commented on src/wallet/wallet.h:
```diff
@@ -351,6 +351,11 @@ class CMerkleTx
     int nIndex;
 
+    /* Refers to height of block against which tx is merkle branch linked to
+     * An block_height == 0 means that tx isn't yet linked to any block
```
The comment says `An block_height == 0` in commit _Add m_block_height field in CMerkleTx_ and then changes to `A block_height == -1` in _Use CWallet::m_last_block_processed_height in GetDepthInMainChain_.

It would be better to not change this in the course of the PR and just set it to `A block_height == -1` in the first PR.

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -1422,12 +1414,13 @@ void CWallet::BlockConnected(const CBlock& block, const std::vector<CTransaction
     // to abandon a transaction and then have it inadvertently cleared by
     // the notification that the conflicted transaction was evicted.
 
+    m_last_block_processed_height += 1;
```
This height accounting in the wallet scares me slightly. I think that prior to this change, the wallet tolerates any bug in the node that sends duplicate `BlockConnected()` or `BlockDisconnected()` calls over the CValidationInterface. After this change, a duplicate `BlockConnected` notification could cause the wallet to get into a very bad inconsistent state.

Is it possible to update the CValidationInterface to also pass the block height when sending BlockConnected/BlockDisconnected notifications?

ariard replied:
You're right, it's far better to rely on message content than counting the message in itself.
To do, I've extended BlockDisconnected, for BlockConnected we have CBlockIndex it was just masked between Notification and CWallet

jnewbery commented:
I've done a quick code review. The main thing that I'd like to see changed is merging the `CMerkleTx` and `CWalletTx` classes now that `CMerkleTx` is accessing wallet data (see discussion here: http://www.erisian.com.au/bitcoin-core-dev/log-2019-07-23.html#l-322)

jnewbery commented:
(excellent work, by the way. Great to see some of these locked_chain instances being cleaned up)

ariard commented on src/wallet/wallet.h:
```diff
@@ -372,6 +372,9 @@ class CMerkleTx
     static const uint256 ABANDON_HASH;
 
+protected:
+    const CWallet* pwallet;
```
See #16451, thanks for going forwad on it :)

Commit:
Pass block height in Chain::BlockConnected/Chain::BlockDisconnected

To do so we update CValidationInterface::BlockDisconnect to take a
CBlockIndex pointing to the block being disconnected.

This new parameter will be use in the following commit to establish
wallet height.

ariard commented:
Addressed @jnewbery comments on a53586c, except https://github.com/bitcoin/bitcoin/pull/15931#discussion_r306961660 where I'm still thinking on it

jnewbery commented:
https://github.com/bitcoin/bitcoin/pull/16451 is merged. You should be able to rebase this and remove one of the commits.

ariard commented:
Rebased

jnewbery commented on src/wallet/wallet.h:
```diff
@@ -831,7 +836,14 @@ class CWallet final : public FillableSigningProvider, private interfaces::Chain:
+     * initialize it to -1.
+     */
+    int m_last_block_processed_height{-1};
+    GUARDED_BY(cs_wallet);
```
I'm not sure if this macro does anything?

ariard replied:
Thanks, I misused some linter scripts I think

jnewbery commented on src/wallet/wallet.h:
```diff
@@ -831,7 +836,14 @@ class CWallet final : public FillableSigningProvider, private interfaces::Chain:
 
     //! Fetches a key from the keypool
-    bool GetKeyFromPool(CPubKey &key, bool internal = false);
+    bool GetKeyFromPool(CPubKey& key, bool internal = false);
```
Please don't fixup style from surrounding code. It makes the diff less clean.

jnewbery reviewed (commented):
just a couple of comments on the first commit

jnewbery commented:
I suggest you reverse the order of the first two commits:

1. have one commit that _only_ updates the signature of `BlockDiscconected` in the validation interface.
2. the second commit adds `m_last_block_processed_height` and uses the `pindex` in `BlockConnected`/`BlockDisconnected` to set the height.

Having one way for setting `m_last_block_processed_height` in the first commit which is replaced by a different way in the second commit makes this confusing for reviewers.

Sjors commented:
Concept ACK on reducing reliance on `locked_chain`. It seems worth adding `m_block_height`.  

I might be more clear to swap the first two (?) commits, thereby introducing `m_last_block_processed_height` without the `+= 1` stuff.

Getting a bunch of warnings on macOS along these lines:

```
In file included from wallet/ismine.cpp:12:
./wallet/wallet.h:846:5: warning: declaration does not declare anything [-Wmissing-declarations]
    GUARDED_BY(cs_wallet);
    ^
./threadsafety.h:18:23: note: expanded from macro 'GUARDED_BY'
#define GUARDED_BY(x) __attribute__((guarded_by(x)))
                      ^
In file included from wallet/ismine.cpp:12:
./wallet/wallet.h:846:5: warning: attribute guarded_by ignored, because it is not attached to a declaration [-Wignored-attributes]
./threadsafety.h:18:48: note: expanded from macro 'GUARDED_BY'
#define GUARDED_BY(x) __attribute__((guarded_by(x)))
                                               ^
2 warnings generated.
```

I'm worried that serializing `m_block_height` in `CWalletTx` makes wallets non backwards compatible, but I can't produce a problem (shameless plug for #12134).

ariard commented:
Thanks for review @Sjors, 

I inverted the 2 top commits at 91dbe47 to ease review.

About `m_block_height` in `CWalletTx`, caching in block height in-memory let us with 2 options at serialization : 
* serialize in-memory member too, deserialize it at wallet loading
* query `getBlockHeight` by passing `hashBlock`at `LoadToWallet` called by `ReadKeyValue`

I tried also to implement 2nd option, but seems to me really inefficient due to the necessity to add another `cs_main` lock in `LoadWallet` for the length of deserialization to preserve locks order.

It may need release notes but I think it's move us towards less interdependence between wallet and chain. What's your thoughts ?

ryanofsky reviewed (commented):
> About `m_block_height` in `CWalletTx`, caching in block height in-memory let us with 2 options at serialization :
> 
> * serialize in-memory member too, deserialize it at wallet loading
> * query `getBlockHeight` by passing `hashBlock`at `LoadToWallet` called by `ReadKeyValue`
> 
> I tried also to implement 2nd option, but seems to me really inefficient due to the necessity to add another `cs_main` lock in `LoadWallet` for the length of deserialization to preserve locks order.
> 
> It may need release notes but I think it's move us towards less interdependence between wallet and chain. What's your thoughts ?

Definitely should go with option #2. (Sorry, I suggested this earlier with https://github.com/bitcoin/bitcoin/pull/15931#discussion_r295434058, but I didn't follow up). Changing wallet serialization makes this PR more complicated to understand and test, and is something that can complicate wallet design going forward by increasing the number of data formats we have to think about staying compatible with.

If there is evidence that 2nd option makes startup time slower (I'd be skeptical because there should be no contention for cs_main and these are in-memory lookups), then serialization could be changed in a more targeted followup PR. It would also be easier to consider other serialization formats in a followup PR, like storing information about blocks separately without changing the transaction format.

jnewbery commented:
> Changing wallet serialization makes this PR more complicated to understand and test, and is something that can complicate wallet design going forward by increasing the number of data formats we have to think about staying compatible with.

Note that `mapValue` can be used to change `WalletTx` serialization in a backwards-compatible way. It's a key-value store that anything can be written to/read from. Take a look at the `ReadOrderPos()`/`WriteOrderPos()` for an example of a parameter being written/read from the `mapValue` store.

@ryanofsky - how can we be sure that if we add a `getBlockHeight()` call for every tx in the wallet, we won't introduce a performance regression for very large wallets?

ryanofsky commented:
> @ryanofsky - how can we be sure that if we add a getBlockHeight() call for every tx in the wallet, we won't introduce a performance regression for very large wallets?

I don't think we can be sure without measuring. I would say if there is a performance regression, I think it would be better to handle in a followup PR that's only changing the serialization and avoiding getBlockHeight() calls, and not changing serialization along with the other things that already have to change here.

The reason I'm skeptical there would be a performance regression is that we're talking about adding a single map lookup in a loop which is already doing multiple allocations, map insertions, disk reads, and deserialization, and with the whole loop preceding a rescan. I wouldn't be shocked to see a difference, but I would be surprised to see a difference, and wouldn't want to make the wallet format more complicated and inconsistent, just for the possibility that a difference might exist.

> Note that mapValue can be used to change WalletTx serialization in a backwards-compatible way. 

Changing the serialization in a backward compatible way would probably be good, but would create some risk for the future (especially without #12134) that we write new code that assumes data is present and seems to work on new wallets, but then turns out not to work on old wallets.

ryanofsky commented:
To be sure, if there is reason to think changing serialization will soup up the wallet, please don't let me stand in the way. My reaction isn't "please don't do this", just "why wouldn't you want to mess with this if you didn't have to?"

ariard commented:
Loading large wallets after initialization will need to lock cs_main to read deserialize their txn, which maybe a performance regression. Removing locks in the wallet would make better but still be a hit.

But agree, this PR is already big enough, I will change for 2nd option, and let's talk about safe serialization in another one, if needed and after more thoughts.

ariard commented:
@ryanofsky in fact I just realized we may have to take 1st option, `bitcoin-wallet` binary load a wallet and can't get its chain interface..

Do you see a 3rd option beyond the 2 I've mentioned ?

ryanofsky commented:
> @ryanofsky in fact I just realized we may have to take 1st option, `bitcoin-wallet` binary load a wallet and can't get its chain interface..
> 
> Do you see a 3rd option beyond the 2 I've mentioned ?

Just skip this if interfaces::Chain* pointer is null. Bitcoin-wallet won't do anything that requires knowing the block height of transactions in the wallet.

Sjors commented:
> Just skip this if interfaces::Chain* pointer is null.

That makes sense, though maybe make that more explicit?

ariard commented:
Yes, I've beefed up comments on this, tell me it's enough (see in `CWallet::Verify` and `CWallet::LoadWallet`)

The query-chain-to-know-height approach makes compatibility not an issue, and doesn't block us to go forward with tx height serialization later . A caveat, right now, it has to lock chain for the whole deserialization period to preserve lock order. This inefficiency should go away when we remove chain locks in wallet (hopefully soon).

Please notify a diff in `AddToWallet` since last review, where we now also check block height of incoming transaction against its old version. Reason is when we load wallet while reindexing, `getBlockHeight` is not going to find block, so even if tx block hash doesn't change, we need to correct its height when we get block notification (`BlockConnected`) after reindex is done.

Also, I didn't pollute wallet db code path by passing the chain lock, as this one may be null in case of wallet tool and lock can be taken recursively without issue. Instead, we check again in `LoadToWallet` with the added method `hasChain` to know if we can query chain state safely.

ariard commented:
Hmm annotated `GetDepthInMainChain` with a `NO_THREAD_SAFETY_ANALYSIS` right now to avoid member access into incomplete type (same as `GetConflicts` and `GetAvailableCredit`). If anyone has an idea how to resolve them smoothly, I will  take it.

jnewbery commented:
I've added a `GetDepthInMainChain()` call back into `SubmitMemoryPoolAndRelay()` in #16557. I think we probably want to get that log fix in first and rebase this PR on top of that (with the _Remove locked_chain from GetDepthInMainChain and its callers_ commit updated to remove _locked_chain_ from `SubmitMemoryPoolAndRelay()`)

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -4628,36 +4651,49 @@ CKeyPool::CKeyPool(const CPubKey& vchPubKeyIn, bool internalIn)
         return 0;
 
-    return locked_chain.getBlockDepth(hashBlock) * (nIndex == -1 ? -1 : 1);
+    // If nIndex is -1 and the block hash is set, it means the transaction is conflicted, and that m_block_height
```
I think an alternative would be to reset the `hashBlock`, `nIndex` and `m_block_height` for all transactions matching the block hash when a `BlockDisconnected` is received. That would involve iterating through all of the wallet txs to check their `hashBlock`s. Re-orgs are supposed to be pretty rare, so I think that's ok. Thoughts?

ariard replied:
Yes I've thought about while working on the rescan logic. I've introduced a `Rewind` callback in the `Chain::Notifications`, it should call a wallet method to zeroed out cleanly tx `hashBlock`, `nIndex` and `m_block_height`. I've proposed to introduce in this future-PR, this one is already wide enough.

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -98,10 +98,7 @@ static std::set<CWallet*> g_unloading_wallet_set;
 // Custom deleter for shared_ptr<CWallet>.
 static void ReleaseWallet(CWallet* wallet)
 {
-    // Unregister and delete the wallet right after BlockUntilSyncedToCurrentChain
```
I don't understand why this has to be removed. Can you add some more detail to the commit log?

(I think the current commit log "Scheduler thread servicing chain state is stopped before wallet unloading and can't be useful" isn't right. `ReleaseWallet()` is the delete expression for the CWallet shared_ptr, and so is called for example when the wallet is unloaded during runtime.)

ariard replied:
I've run tests again, can't remember but I had a test staling some point, I think I may have introduce a height error at some point which would false `waitForNotificationsIfTipChanged` and later corrected it without realizing it would fix this. Drop this commit.

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -1332,12 +1329,12 @@ bool CWallet::AbandonTransaction(interfaces::Chain::Lock& locked_chain, const ui
     LOCK(cs_wallet);
 
-    int conflictconfirms = -locked_chain->getBlockDepth(hashBlock);
+    int conflictconfirms = -conflicting_height;
```
I think this is a change in behaviour. The was previously (current tip height) - (height of block with conflicting tx). This is now just (height of block with conflicting tx).

ariard replied:
Thanks for this one, in fact after thinking of it, that would be a bug if you have multiple conflicting txn, you would pick up the one with the highest height instead of the highest depth so not the most conflicting one..

jnewbery reviewed (commented):
I really like this PR. I've reviewed all the commits and left some inline comments. One other suggested change:

The _Restrain waitForNotificationsIfNewBlocksConnected check to strict tip_ commit log could be a bit clearer. I suggest something like:

> `waitForNotificationsIfNewBlocksConnected()` would previously return immediately if the best block processed by the wallet was a descendant of the node's tip. This would mean that during a reorg, an RPC could return before the `BlockDisconnected` callback had been triggered. This could cause a tx that had been included in the disconnected block to be displayed as confirmed, for example.
> 
> Instead, change `waitForNotificationsIfNewBlocksConnected()` to `waitForNotificationsIfTipChanged()` and only return early if the best block processed by the wallet is the node's tip.

I'd also recommend updating the text in the PR description.

ariard commented on src/wallet/wallet.h:
```diff
@@ -833,6 +845,12 @@ class CWallet final : public FillableSigningProvider, private interfaces::Chain:
+     * without relying on Chain interface beyond asynchronous updates. For safety, we
+     * initialize it to -1.
+     */
+    int m_last_block_processed_height GUARDED_BY(cs_wallet) = -1;
```
Yes, had a comment about it, it doesn't imply that the wallet has seen all blocks since genesis up until `m_last_block_processed_height`

ariard replied:
Updated comments

ariard commented on src/wallet/wallet.cpp:
```diff
@@ -1118,16 +1115,12 @@ bool CWallet::AddToWallet(const CWalletTx& wtxIn, bool fFlushOnClose)
-        // If no longer abandoned, update
-        if (wtxIn.hashBlock.IsNull() && wtx.isAbandoned())
-        {
+        assert(!wtxIn.isAbandoned()); // Incoming transactions should never have special abandoned block hash
```
After reading again about how to use abandontransaction, I think you're right, tx might be not in a block or not in your mempool so you can flag is as abandoned. But after this, it can still being included as part of other peers mempools. Removed it, was suggested here initially https://github.com/bitcoin/bitcoin/pull/15931#discussion_r295463739

ariard commented on src/wallet/wallet.cpp:
```diff
@@ -1180,9 +1173,14 @@ bool CWallet::AddToWallet(const CWalletTx& wtxIn, bool fFlushOnClose)
 {
     uint256 hash = wtxIn.GetHash();
+    Optional<int> block_height = MakeOptional(false, int());
+    if (hasChain()) {
```
Okay, this point is already mentioned in commit message, `Verify` and `LoadWallet` but added also on this place.

jkczyz replied:
Instead of initializing an `Optional<int>` then conditionally assigning it based on the result of `hasChain()`, could you instead define something like `chainHeight()` to return an `Optional<int>`? This would hide the chain logic from `LoadToWallet()`.

If not, consider directly assigning `0` to `wtxIn.m_block_height` and conditionally updating it based on `hasChain()`. The addition of `Optional<int>` here just seems to add indirection for the reader.

ariard replied:
I don't want to add new method in Chain interface, but yes 2nd suggested alternative avoid the `MakeOptional` which is good, updated

ariard commented on src/wallet/wallet.cpp:
```diff
@@ -1248,9 +1246,7 @@ bool CWallet::AddToWalletIfInvolvingMe(const CTransactionRef& ptx, const uint256
             CWalletTx wtx(this, ptx);
 
-            // Get merkle branch if transaction was found in a block
-            if (!block_hash.IsNull())
```
Gone this way, see https://github.com/bitcoin/bitcoin/pull/16624

promag commented on src/validationinterface.h:
```diff
@@ -114,7 +114,7 @@ class CValidationInterface {
      * Called on a background thread.
      */
-    virtual void BlockDisconnected(const std::shared_ptr<const CBlock> &block) {}
+    virtual void BlockDisconnected(const std::shared_ptr<const CBlock>& block, const CBlockIndex* pindex) {}
```
3a04cd6776f7f1ec86cad20de90c9fbb24655536

nit, could drop `&` from 1st argument, it was discussed elsewhere that smart pointers and iterators should be used as value.

jnewbery replied:
Please don't do this. Trying to 'fix' surrounding code in a PR makes the diff bigger and more confusing for reviewers.

promag commented on src/wallet/wallet.cpp:
```diff
@@ -1442,9 +1443,12 @@ void CWallet::BlockDisconnected(const CBlock& block, int height)
         SyncTransaction(ptx, {} /* block hash */, 0 /* position in block */);
     }
+
+    m_last_block_processed = block.hashPrevBlock;
```
3d32a691d2df4c73a39a64a5d331c29b5eb94213

Is this a bugfix?

ariard replied:
Yes it was a bugfix compare to a53586c at least, it become obvious only after switching to query the chain based on block hash at wallet loading!

promag commented on src/wallet/wallet.h:
```diff
@@ -1356,6 +1363,19 @@ class CWallet final : public FillableSigningProvider, private interfaces::Chain:
+        return m_last_block_processed_height;
+    };
+    /** Set last block processed height, currently only use in unit test */
+    void SetLastBlockHeight(int block_height) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
```
3d32a691d2df4c73a39a64a5d331c29b5eb94213

Why isn't this named `SetLastBlockProcessedHeight`?

Anyway I think the variable should be `m_chain_tip_height`. Also, I think now `m_last_block_processed` should be improved, like:
```cpp
    uint256 m_chain_tip_hash GUARDED_BY(cs_wallet);
    int m_chain_tip_height GUARDED_BY(cs_wallet){-1};
```
and both should be set at once.

ariard replied:
Agree for the name of method and set at once, but not sure for variables. I think it would be confusing, as intent of this PR is to let the wallet have its own view of what chain tip is after processing block from validation interface. So real chain tip and the wallet one may diverge.

promag replied:
Yeah these would be from the wallet POV.

jkczyz commented on src/wallet/wallet.h:
```diff
@@ -488,6 +489,11 @@ class CWalletTx
      */
     int nIndex;
 
+    /* Refers to height of block against which tx is merkle branch linked to
```
Nits:

1. Consider rewording for succinctness and clarity: Height of block against which tx is linked via a merkle branch.
2. End each sentence with a period as they tend to blend when reading the comment.

ariard commented on src/wallet/wallet.cpp:
```diff
@@ -1425,6 +1425,7 @@ void CWallet::BlockConnected(const CBlock& block, const std::vector<CTransaction
     // to abandon a transaction and then have it inadvertently cleared by
     // the notification that the conflicted transaction was evicted.
 
+    m_last_block_processed_height = height;
```
Yes it has to be before, but IIRC not the m_last_block_processed assignment. Moved it near to m_last_block_processed_height.

ariard commented:
Updated at f766fe4, including @promag and @jkczyz feedbacks.

ariard commented:
Rebased at 7427cbf, due to silent build failures not reported by merge conflict detection.

dergigi commented:
Tried to compile commit `7427cbf7a` to prepare for the PR review club, and I'm getting an error when running `make`:

```
  CXX      qt/test/test_bitcoin_qt-wallettests.o
qt/test/wallettests.cpp: In function ‘void {anonymous}::TestGUI()’:
qt/test/wallettests.cpp:145:17: error: ‘using element_type = class CWallet’ {aka ‘class CWallet’} has no member named ‘SetLastBlockProcessedHeight’; did you mean ‘SetLastBlockProcessed’?
  145 |         wallet->SetLastBlockProcessedHeight(105, ::ChainActive().Tip()->GetBlockHash());
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                 SetLastBlockProcessed
make[2]: *** [Makefile:11357: qt/test/test_bitcoin_qt-wallettests.o] Error 1
make[2]: *** [Makefile:11357: qt/test/test_bitcoin_qt-wallettests.o] Error 1
make[1]: *** [Makefile:13676: all-recursive] Error 1
make: *** [Makefile:774: all-recursive] Error 1
```

Not sure if I'm doing anything wrong. I am able to compile `master` without any errors.

ariard commented:
@dergigi sorry for that, skipped qt test build when I replied promag comment yesterday, shouldn't have. Updated at 15100ee

jonatack commented on src/validationinterface.cpp:
```diff
@@ -31,7 +31,7 @@ struct MainSignalsInstance {
     boost::signals2::signal<void (const CTransactionRef &)> TransactionAddedToMempool;
     boost::signals2::signal<void (const std::shared_ptr<const CBlock> &, const CBlockIndex *pindex, const std::vector<CTransactionRef>&)> BlockConnected;
-    boost::signals2::signal<void (const std::shared_ptr<const CBlock> &)> BlockDisconnected;
+    boost::signals2::signal<void(const std::shared_ptr<const CBlock>&, const CBlockIndex* pindex)> BlockDisconnected;
```
nit: missing space between `void` and `(const` if you have to retouch this PR

ryanofsky commented:
I was looking over this again and think it's a really nice change. It'd be good to rebase now that #16624 is merged. It would also help with #16426 which depends on this PR.

ariard commented:
Thanks for the push @ryanofsky, and sorry for the delay for following-up on this PR. Also added nits leftover from #16624 

I spent a bit more time thinking how we handle conflicts and found that Confirmation status isn't correct for children of conflicted txn, if conflicting tx get removed we unconfirm well the parent conflicted tx but not its child. Test here : https://github.com/ariard/bitcoin/tree/2019-10-conflicted-txn,  will correct in a later PR to avoid delaying more this one.

ryanofsky reviewed (approved):
Light code review ACK 4f5d03a4c68e0576a85add4254f4f4470897a0fa. I reviewed the whole thing and think it looks good, but did have some questions below.

re: https://github.com/bitcoin/bitcoin/pull/15931#issuecomment-541514108

> found that Confirmation status isn't correct for children of conflicted txn

IIUC this is current behavior, independent of the PR?

ariard commented on src/wallet/wallet.cpp:
```diff
@@ -1168,13 +1170,17 @@ void CWallet::LoadToWallet(CWalletTx& wtxIn)
+            wtxIn.m_confirm.block_height = 0;
             wtxIn.m_confirm.nIndex = 0;
+        } else {
+            wtxIn.m_confirm.block_height = (block_height) ? *block_height : 0;
```
Updated the comment, which was saying the inverse of what was done in code, if tx was confirmed (or conflicted by another confirmed tx) and a reorg happens we should reset tx status to unconfirmed. The case where a transaction was reorged out while online and then reconfirmed while offline is covered by the rescan logic of wallet creation IMO. Reason for the `&& !block_height` is to check that we don't find anymore the block based on the tx `hashBlock` stored, hint of a reorg.

ariard commented on src/wallet/wallet.h:
```diff
@@ -976,6 +978,8 @@ class CWallet final : public FillableSigningProvider, private interfaces::Chain:
     interfaces::Chain& chain() const { assert(m_chain); return *m_chain; }
 
+    /** Interface to assert chain access. */
+    bool hasChain() const { return m_chain != nullptr; }
```
Removed

ariard commented on src/interfaces/chain.cpp:
```diff
@@ -348,13 +348,11 @@ class ChainImpl : public Chain
         if (!old_tip.IsNull()) {
             LOCK(::cs_main);
             if (old_tip == ::ChainActive().Tip()->GetBlockHash()) return;
-            CBlockIndex* block = LookupBlockIndex(old_tip);
```
Updated commit message, due to the switch to height tracking based on callbacks, if we don't wait for the whole callbacks queue being drawn because the last block hash processed by the wallet may be a descendant of the current tip, we may miss to process BlockDisconnected and so get tx deptch inconsistencies.

ariard commented on src/wallet/wallet.cpp:
```diff
@@ -4669,10 +4669,13 @@ void CWalletTx::SetConf(Status status, const uint256& block_hash, int posInBlock
 int CWalletTx::GetDepthInMainChain(interfaces::Chain::Lock& locked_chain) const
 {
+    if (pwallet == nullptr)
+        return 0;
```
Switch to asserting.

ariard commented on src/wallet/wallet.h:
```diff
@@ -639,7 +639,13 @@ class CWalletTx
+    // resolve the issue of member access into incomplete type CWallet. Note
+    // that we still have the runtime check "AssertLockHeld(pwallet->cs_wallet)"
+    // in place.
+    int GetDepthInMainChain(interfaces::Chain::Lock& locked_chain) const NO_THREAD_SAFETY_ANALYSIS;
```
Ok, I think will let it as it is for now and will try to pass a `CWallet&` argment in a future PR scoped only on this.

ariard commented:
Thanks for review @ryanofsky, rebased and addressed your comment at 089189e

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -1391,7 +1398,7 @@ void CWallet::SyncTransaction(const CTransactionRef& ptx, CWalletTx::Status stat
     auto locked_chain = chain().lock();
     LOCK(cs_wallet);
-    SyncTransaction(ptx, CWalletTx::Status::UNCONFIRMED, {} /* block hash */, 0 /* position in block */);
+    SyncTransaction(ptx, CWalletTx::Status::UNCONFIRMED, {} /* block hash */, 0 /* position in block */, 0 /* block height */);
```
In commit "Pass block height in Chain::BlockConnected/Chain::BlockDisconnected" (82c591ea35ac918147ca66444ae608b8385abd7c)

Order of arguments here is wrong. Block height come before position in block.

ariard replied:
Added commit d0c4ce6 : Replace CWalletTx:SetConf by Confirmation initialization list, should make things less error-prone.

ryanofsky commented on src/wallet/wallet.h:
```diff
@@ -798,10 +800,10 @@ class CWallet final : public FillableSigningProvider, private interfaces::Chain:
      * posInBlock signals or by checking mempool presence when necessary.
      */
-    bool AddToWalletIfInvolvingMe(const CTransactionRef& tx, CWalletTx::Status status, const uint256& block_hash, int posInBlock, bool fUpdate) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
+    bool AddToWalletIfInvolvingMe(const CTransactionRef& tx, CWalletTx::Status status, const uint256& block_hash, int posInBlock, bool fUpdate, int block_height) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
```
In commit "Add block_height field in struct Confirmation" (de8fe7d717042823f2aa429be2a260904a10b42e)

Could you move the block_height argument after block_hash to be consistent with `SyncTransaction` argument order and `Confirmation` struct order?

Alternately, it possible, it would be really nice to combine the 4 status/hash/height/index arguments into a single CWalletTx::Confirmation struct argument, because having 4 int arguments in a row makes bugs more likely.

ariard replied:
See d0c4ce6

ryanofsky commented on src/wallet/wallet.h:
```diff
@@ -630,7 +630,7 @@ class CWalletTx
     std::set<uint256> GetConflicts() const NO_THREAD_SAFETY_ANALYSIS;
 
-    void SetConf(Status status, const uint256& block_hash, int posInBlock);
+    void SetConf(Status status, const uint256& block_hash, int posInBlock, int block_height);
```
In commit "Add block_height field in struct Confirmation" (de8fe7d717042823f2aa429be2a260904a10b42e)

Could you move the block_height argument after block_hash to be consistent with `SyncTransaction` argument order and `Confirmation` struct order?

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -4632,20 +4639,24 @@ CKeyPool::CKeyPool(const CPubKey& vchPubKeyIn, bool internalIn)
 
 int CWalletTx::GetDepthInMainChain(interfaces::Chain::Lock& locked_chain) const
 {
+    assert(m_confirm.block_height >= -1);
```
In commit "Add block_height field in struct Confirmation" (de8fe7d717042823f2aa429be2a260904a10b42e)

This seems harmless, but it's unclear why it's part of this commit. Is this line meant to added to another commit where the block_height is actually used? Also, I don't see where -1 would be set or allowed for this.

ariard replied:
Removed, was there to be sure block_height was initialized in a previous version, now it has to be with `Confirmation` default initialization

ryanofsky commented on src/wallet/wallet.h:
```diff
@@ -496,13 +496,14 @@ class CWalletTx
-     * This pair is both 0 if tx hasn't confirmed yet. Meaning of these fields changes with CONFLICTED state
-     * where they instead point to block hash and index of the deepest conflicting tx.
+    /* Confirmation includes tx status and a triplet of {block hash/block height/tx index in block} at which tx has been confirmed.
+     * This triplet is whole 0 if tx hasn't confirmed yet. Meaning of these fields changes with CONFLICTED state
```
In commit "Add block_height field in struct Confirmation" (de8fe7d717042823f2aa429be2a260904a10b42e)

Maybe replace "hasn't confirmed yet" with "is unconfirmed or abandoned" to be more specific and include mention of the abandoned state.

ryanofsky commented on src/wallet/wallet.h:
```diff
@@ -546,7 +547,6 @@ class CWalletTx
          * compatibility (pre-commit 9ac63d6).
          */
         if (serializedIndex == -1 && m_confirm.hashBlock == ABANDON_HASH) {
-            m_confirm.hashBlock = uint256();
```
In commit "Add block_height field in struct Confirmation" (de8fe7d717042823f2aa429be2a260904a10b42e)

Note: this is another cleanup from the previous pr.

ryanofsky commented on src/wallet/wallet.h:
```diff
@@ -1405,6 +1411,20 @@ class CWallet final : public FillableSigningProvider, private interfaces::Chain:
+    int GetLastBlockHeight() const EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
+    {
+        AssertLockHeld(cs_wallet);
+        return m_last_block_processed_height;
```
In commit "Add m_last_block_processed_height field in CWallet" (9161f0bf0321e90e0aaeeafc58c5ea6f97ebb1e2):

Would `assert(m_last_block_processed_height >= 0)` to make sure we're past initialization and don't wind up with crazy depth values.

ariard replied:
Done

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -1416,7 +1423,7 @@ void CWallet::BlockConnected(const CBlock& block, const std::vector<CTransaction
     m_last_block_processed = block_hash;
     for (size_t i = 0; i < block.vtx.size(); i++) {
-        SyncTransaction(block.vtx[i], CWalletTx::Status::CONFIRMED, block_hash, i);
+        SyncTransaction(block.vtx[i], CWalletTx::Status::CONFIRMED, block_hash, m_last_block_processed_height, i);
```
In commit "Add block_height field in struct Confirmation" (de8fe7d717042823f2aa429be2a260904a10b42e)

Can you replace m_last_block_processed_height with height on this line? It's slightly worrying to depend on the wallet member if it might get updated for some reason, and also better to be more consistent with previous `block_hash` argument.

ryanofsky reviewed (commented):
Started closer review (will update list below with progress).
- [X] 82c591ea35ac918147ca66444ae608b8385abd7c Pass block height in Chain::BlockConnected/Chain::BlockDisconnected (1/7)
- [X] 9161f0bf0321e90e0aaeeafc58c5ea6f97ebb1e2 Add m_last_block_processed_height field in CWallet (2/7)
- [X] de8fe7d717042823f2aa429be2a260904a10b42e Add block_height field in struct Confirmation (3/7)
- [x] acae8e41fbcec5ee969d511e736e9238970aa087 Restrain waitForNotificationsIfNewBlocksConnected check to strict tip (4/7)
- [x] 8033be2b111af4c741d5c0ea715381976abd4c15 Use CWallet::m_last_block_processed_height in GetDepthInMainChain (5/7)
- [x] 360237eaa9e4ad11865c00407a4ef444df841923 Remove locked_chain from GetDepthInMainChain and its callers (6/7)
- [x] 089189ef99effe79bb2b71b9248f8aa419079b9a Remove getBlockDepth method from Chain::interface (7/7)

Commit:
Replace CWalletTx::SetConf by Confirmation initialization list

ryanofsky reviewed (approved):
Code review ACK 089189ef99effe79bb2b71b9248f8aa419079b9a. I finished reviewing the last few commits, which were all straightforward, and I didn't see anything I'd change about them. I forgot to mention this yesterday, but please ignore my review suggestions from then if they seem like too much work, or not worth the effort. I think this change should basically be ready to merge in its current form after a rebase and a few re-acks.

ariard commented on src/wallet/wallet.h:
```diff
@@ -496,13 +496,14 @@ class CWalletTx
     struct Confirmation {
         Status status = UNCONFIRMED;
         uint256 hashBlock = uint256();
+        int block_height = 0;
```
I let it be, need thoughts maybe to retrieve more as same time

ariard commented:
Thanks @ryanofsky , addressed most of your comments, specially https://github.com/bitcoin/bitcoin/pull/15931#discussion_r338169478a and adding commit d0c4ce6 to wrap all confirmations values directly in BlockConnected/BlockDisconnected.

Rebased at 75c276e

ryanofsky commented on src/wallet/rpcdump.cpp:
```diff
@@ -383,7 +383,8 @@ UniValue importprunedfunds(const JSONRPCRequest& request)
 
-    wtx.SetConf(CWalletTx::Status::CONFIRMED, merkleBlock.header.GetHash(), txnIndex);
+    CWalletTx::Confirmation confirm(CWalletTx::Status::CONFIRMED, merkleBlock.header.GetHash(), txnIndex);
+    wtx.m_confirm = confirm;
```
In commit "Replace CWalletTx::SetConf by Confirmation initialization list" (d0c4ce686ba388b9fefff2b153358e33a042ee80)

Not important, but might want to avoid creating named variable with:

```
wtx.m_confirm = CWalletTx::Confirmation{CWalletTx::Status::CONFIRMED, merkleBlock.header.GetHash(), txnIndex};
```

Similarly with all the other temporary `confirm` variables below.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -1193,19 +1193,19 @@ void CWallet::LoadToWallet(CWalletTx& wtxIn)
 }
 
-bool CWallet::AddToWalletIfInvolvingMe(const CTransactionRef& ptx, CWalletTx::Status status, const uint256& block_hash, int posInBlock, bool fUpdate)
+bool CWallet::AddToWalletIfInvolvingMe(const CTransactionRef& ptx, CWalletTx::Confirmation confirm, bool fUpdate)
```
In commit "Replace CWalletTx::SetConf by Confirmation initialization list" (d0c4ce686ba388b9fefff2b153358e33a042ee80)

Not a big deal, but probably better to pass with reference type `const CWalletTx::Confirmation&` to avoid copies when this is called. Similarly for other functions taking confirm arguments.

ryanofsky reviewed (approved):
Code review ACK 75c276ea9bcb95f549f872ad365313ec7bd76da3. Changes since last review: rebase, space formatting stuff, requested new assert in GetLastBlockHeight, new commit passing Confirm structs instead of individual properties, and various other requested changes. Looks very good. I left two suggestions for the new commit, but feel free to ignore

laanwj commented:
Concept ACK. I like the direction of this.

ryanofsky reviewed (approved):
Code review ACK b2c68eb3cdc851d4d1a3bb6a8a64397907e22618. No actual changes since last review, just rebase due to adjacent conflicts with #17260

jonatack commented:
Code review in progress. Built b2c68eb3cdc851d4d1a3bb6a8a64397907e22618, ran tests, running bitcoind w/logging for a few hours.

ryanofsky commented:
This has had a decent amount of code review but only one code review ack.

If previous reviewers want to take another look and add ACKs, it would be very helpful. The PR is smaller and simpler than it was originally, and most of the changes (especially in the last few commits) are actually trivial. A lot of improvments can build on this PR to make the wallet more asynchronous, clean up rescanning, etc.

Review history:

ryanofsky review and concept ack 2019-06-19 https://github.com/bitcoin/bitcoin/pull/15931#pullrequestreview-251859358
promag concept ack 2019-06-30 https://github.com/bitcoin/bitcoin/pull/15931#pullrequestreview-256075977
jnewbery review 2019-07-23 https://github.com/bitcoin/bitcoin/pull/15931#pullrequestreview-265517375
jnewbery review 2019-08-06 https://github.com/bitcoin/bitcoin/pull/15931#pullrequestreview-271423875
sjors review and concept ack 2019-08-06 https://github.com/bitcoin/bitcoin/pull/15931#issuecomment-518725058
jnewbery review 2019-08-09 https://github.com/bitcoin/bitcoin/pull/15931#pullrequestreview-273228460
promag review 2019-08-15 https://github.com/bitcoin/bitcoin/pull/15931#pullrequestreview-275692610
jkczyz review 2019-08-19 https://github.com/bitcoin/bitcoin/pull/15931#pullrequestreview-276662148
ryanofsky review 2019-10-23 https://github.com/bitcoin/bitcoin/pull/15931#pullrequestreview-306024592
ryanofsky review ack 2019-10-24 https://github.com/bitcoin/bitcoin/pull/15931#pullrequestreview-306864462
laanwj concept ack 2019-10-25 https://github.com/bitcoin/bitcoin/pull/15931#issuecomment-546341167
ryanofsky reack 2019-10-30 https://github.com/bitcoin/bitcoin/pull/15931#pullrequestreview-309395041

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -1840,7 +1860,7 @@ bool CWalletTx::IsTrusted(interfaces::Chain::Lock& locked_chain) const
     if (!locked_chain.checkFinalTx(*tx)) {
```
Note for future followup: I believe we could call [`IsFinalTx`](https://github.com/bitcoin/bitcoin/blob/50591f6ec61b802cf4193cdbefcc85ad75716e8d/src/consensus/tx_verify.cpp#L17-L28) instead of [`checkFinalTx`](https://github.com/bitcoin/bitcoin/blob/50591f6ec61b802cf4193cdbefcc85ad75716e8d/src/validation.cpp#L189-L219) so `locked_chain` argument no longer needs to be passed to `CWalletTx::IsTrusted`

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1598,7 +1597,7 @@ static UniValue listsinceblock(const JSONRPCRequest& request)
     for (const std::pair<const uint256, CWalletTx>& pairWtx : pwallet->mapWallet) {
         CWalletTx tx = pairWtx.second;
 
-        if (depth == -1 || abs(tx.GetDepthInMainChain(*locked_chain)) < depth) {
```
In commit "Remove locked_chain from GetDepthInMainChain and its callers" (dd3cb3804126abdba8684cb965d2e1dc3766b3fd)

Appears to be a mistake during rebase: lost abs() call

ariard replied:
Thanks to spot it, was introducted in 436ad43, corrected rebase mistake!

ryanofsky reviewed (approved):
Code review ACK 8a11ab3ff4cba6456021e46813f530c22ada8ac1. No changes since previous review other than rebase

promag commented:
Sure Russell, I'll review.

jnewbery reviewed (commented):
utACK 8a11ab3ff4cba6456021e46813f530c22ada8ac1

I have a couple of nits inline, which can be fixed up later in order to not invalidate review. There are a couple of changes that I would like fixed before merge, since they're to do with the git history, so can't be fixed after the fact (and they don't change the final code, so should be an easy re-ACK for other reviewers)

1. Add the `block_height` member in the _Add block_height field in struct Confirmation_ to not potentially break a git bisect and make the history correct.
2. The wording for commit _Restrain waitForNotificationsIfNewBlocksConnected check to strict tip_ is a little confusing. Can I suggest the following instead:

```
Only return early from BlockUntilSyncedToCurrentChain() if current tip is exact match

In the next commit, we start using BlockConnected/BlockDisconnected
callbacks to establish tx depth, rather than querying the chain
directly.

Currently, BlockUntilSyncedToCurrentChain() will return early if the
best block processed by the wallet is a descendant of the node's tip.
That means that in the case of a re-org, it won't wait for the
BlockDisconnected callbacks that have been enqueued during the re-org
but have not yet been triggered in the wallet.

Change BlockUntilSyncedToCurrentChain() to only return early if the
wallet's m_last_block_processed matches the tip exactly. This ensures
that there are no BlockDisconnected or BlockConnected callbacks
in-flight.
```

Great work with this PR. Thanks for persisting with it, and thanks to @ryanofsky for chasing reviewers!

jkczyz reviewed (commented):
utACK 8a11ab3 modulo a few comments.

Commit:
Refactor some importprunedfunds checks with guard clause

Credit to jkczyz

ariard commented:
Updated at 9810862 :
* move `block_height` to _Add block_height field in struct Confirmation_
* modify `Confirmation` constructor and add comments to avoid ambiguity in place where it's called
* rearrange `Confirmation` constructor comment
* fix nit wording on `Confirmation` comment
* add guard clause for `importprunedfunds` as [suggested](https://github.com/bitcoin/bitcoin/pull/15931#discussion_r342821318) in its own commit
* rewrite commit message for former _Restrain waitForNotificationsIfNewBlocksConnected check to strict tip_

Thanks @ryanofsky, @jnewbery, @jkczyz for reviews! Re-ACK should be minimal on diff.

ariard commented on src/wallet/wallet.h:
```diff
@@ -356,14 +356,16 @@ class CWalletTx
+        uint256 hashBlock;
+        int block_height;
+        int nIndex;
+        Confirmation(Status s = UNCONFIRMED, uint256 h = uint256(), int b = 0, int i = 0) : status(s), hashBlock(h), block_height(b), nIndex(i) {}
```
Rearrange members initialization order and add comments, tell me if you have a better measure.

ariard commented on src/wallet/wallet.h:
```diff
@@ -361,9 +361,11 @@ class CWalletTx
+        Status status;
+        uint256 hashBlock;
+        int block_height;
+        int nIndex;
```
I think we can't due to setting them with method like `setAbandoned`?

jkczyz replied:
Ah, you're right. Could be better to assign a new object within `setAbandoned` instead, but feel free to leave as it is. Looks like there are other places where these are directly assigned.

ariard commented on src/wallet/wallet.h:
```diff
@@ -361,9 +361,11 @@ class CWalletTx
-        int nIndex = 0;
+        Status status;
+        uint256 hashBlock;
+        int block_height;
```
Shouldn't be an issue anymore as `block_height` has been switched to the right commit.

ryanofsky reviewed (approved):
Code review ACK 36b68de5b2938722911db900ca299f7008780d01. Changes since last review: new jkczyz refactor importprunedfunds commit, changed BlockUntilSyncedToCurrentChainChanges commit title and description, changed Confirmation struct field order and line-wrapped comment

jnewbery commented on src/wallet/wallet.h:
```diff
@@ -356,14 +356,17 @@ class CWalletTx
+        int block_height;
+        uint256 hashBlock;
+        int nIndex;
+        Confirmation(Status s = UNCONFIRMED, int b = 0, uint256 h = uint256(), int i = 0) : status(s), block_height(b), hashBlock(h), nIndex(i) {}
```
very minor nit: using `b` for height and `h` for hash is a little confusing. Using a few extra characters and calling these `status_in`, `height_in`, `hash_in` and `index_in` would make this clearer.

jnewbery reviewed (commented):
utACK 36b68de5b2938722911db900ca299f7008780d01

I've left one nit, which you definitely shouldn't fix unless you need to change something else.

jkczyz commented:
utACK 769ff05

jnewbery commented:
@jkczyz you've ACKed an intermediate commit (github annoyingly orders commits in date order, not commit order). Did you mean to ACK the final commit in this branch (36b68de5b2938722911db900ca299f7008780d01).

jkczyz commented:
> @jkczyz you've ACKed an intermediate commit (github annoyingly orders commits in date order, not commit order). Did you mean to ACK the final commit in this branch ([36b68de](https://github.com/bitcoin/bitcoin/commit/36b68de5b2938722911db900ca299f7008780d01)).

utACK 36b68de

Right! The PR interface was showing a different order from the branch.

promag commented on src/wallet/wallet.cpp:
```diff
@@ -3936,9 +3936,11 @@ CKeyPool::CKeyPool(const CPubKey& vchPubKeyIn, bool internalIn)
     if (isUnconfirmed() || isAbandoned()) return 0;
 
-    return locked_chain.getBlockDepth(m_confirm.hashBlock) * (isConflicted() ? -1 : 1);
+    return (pwallet->GetLastBlockHeight() - m_confirm.block_height + 1) * (isConflicted() ? -1 : 1);
```
0ff03871add000f8b4d8f82aeb168eed2fc9dc5f

I don't think commit message is correct:
```
Avoid to lock chain to query state thanks to tracking last block
height in CWallet.
```
Should say something like?
```
Use wallet's last block height instead of the locked chain, which
will allow removing the chain lock.
```

ariard replied:
That's a better wording, I will use it if I have to rebase.

promag commented on src/wallet/wallet.cpp:
```diff
@@ -986,7 +986,7 @@ void CWallet::MarkConflicted(const uint256& hashBlock, int conflicting_height, c
     LOCK(cs_wallet);
 
-    int conflictconfirms = -locked_chain->getBlockDepth(hashBlock);
+    int conflictconfirms = (m_last_block_processed_height - conflicting_height + 1) * -1;
```
36b68de5b2938722911db900ca299f7008780d01

What's the problem with `-(m_last_block_processed_height - conflicting_height + 1)`?

promag reviewed (commented):
Code review ACK 36b68de5b2938722911db900ca299f7008780d01.

Build some intermediate commits too. Some comments for your consideration.

ariard commented on src/qt/test/wallettests.cpp:
```diff
@@ -139,10 +139,12 @@ void TestGUI(interfaces::Node& node)
     wallet->LoadWallet(firstRun);
     {
         auto spk_man = wallet->GetLegacyScriptPubKeyMan();
+        auto locked_chain = wallet->chain().lock();
```
Method `chain` here is the way to use the `interfaces::chain`, I think it doesn't matter that much here to use the `Node` or `Chain` one.

ariard commented:
Can someome relaunch AppVeyor ? It got a linker error on bech32_tests.obj

jonatack commented:
> Can someome relaunch AppVeyor ? It got a linker error on bech32_tests.obj

Appveyor is an unrelated pending issue from https://github.com/bitcoin/bitcoin/pull/17357 that should be fixed by https://github.com/bitcoin/bitcoin/pull/17384.

meshcollider commented:
Looks good, thanks for keeping this maintained ariard. I agree with most of promag's comments but they aren't worth holding up merge for.

utACK 36b68de5b2938722911db900ca299f7008780d01

MarcoFalke commented:
It looks like this might have sped up all WalletBalance micro benchmarks:

https://codespeed.bitcoinperf.com/timeline/?exe=3%2C4%2C2%2C1%2C5&base=1%2B23&ben=micro.clang.WalletBalanceDirty&env=1&revs=10&equid=off&quarts=on&extr=on

shesek commented:
It appears like @MarcoFalke's link always displays the last 10 days, which already don't include the performance gains (which happened on Nov 8). I couldn't find a way to link to a specific date range in codespeed, so for future reference for people bumping into this thread, here's [a screenshot](https://user-images.githubusercontent.com/877904/69193689-fdf0b300-0b2f-11ea-9d1d-5ab40ef5cf00.png) instead (of the last 50 days).

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -3936,9 +3936,11 @@ CKeyPool::CKeyPool(const CPubKey& vchPubKeyIn, bool internalIn)
     if (isUnconfirmed() || isAbandoned()) return 0;
 
-    return locked_chain.getBlockDepth(m_confirm.hashBlock) * (isConflicted() ? -1 : 1);
+    return (pwallet->GetLastBlockHeight() - m_confirm.block_height + 1) * (isConflicted() ? -1 : 1);
```
In commit "Use CWallet::m_last_block_processed_height in GetDepthInMainChain" (0ff03871add000f8b4d8f82aeb168eed2fc9dc5f)

@ariard It seems like this line is written assuming that `GetLastBlockHeight()` is always `>= m_confirm.block_height`, but if I add asserts here:

```c++
assert(m_confirm.block_height >= 0);
assert(m_confirm.block_height <= pwallet->GetLastBlockHeight());
```

The second assert fails in `wallet_abandonconflict.py` and `wallet_bumpfee.py` tests. Wondering if this expected or seems like a bug to you

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -3936,9 +3936,11 @@ CKeyPool::CKeyPool(const CPubKey& vchPubKeyIn, bool internalIn)
+    AssertLockHeld(pwallet->cs_wallet);
     if (isUnconfirmed() || isAbandoned()) return 0;
 
-    return locked_chain.getBlockDepth(m_confirm.hashBlock) * (isConflicted() ? -1 : 1);
```
In commit "Use CWallet::m_last_block_processed_height in GetDepthInMainChain" ([0ff0387](https://github.com/bitcoin/bitcoin/commit/0ff03871add000f8b4d8f82aeb168eed2fc9dc5f))

Dropping the `getBlockDepth` call here introduces a bug because `getBlockDepth` had a safety check to return 0 if there was a reorg and hashBlock is no longer in the active chain, while the new code will just return a garbage depth. I think this is also the reason adding height asserts didn't work in some tests: https://github.com/bitcoin/bitcoin/pull/15931#discussion_r371512762. A few fixes are possible and mentioned [*bug-reorg-depth*](https://github.com/bitcoin-core/bitcoin-devwiki/wiki/Wallet-Transaction-Conflict-Tracking#bug-reorg-depth). A fix shouldn't be hard, but a bigger challenge is adding better test coverage to prevent more bugs like this.
